enzyme/Enzyme/MLIR/Analysis/ActivityAnalysis.cpp

```diff
@@ -0,0 +1,3191 @@
+//       InsertConstantInstruction(TR, I);
+//       return true;
+
+//     default:
+//       break;
+//     }
+//   }
+
+  // Analyzer for inductive assumption where we attempt to prove this is
+  // inactive from a lack of active users
+  std::shared_ptr<ActivityAnalyzer> DownHypothesis;
+
+  // If this instruction does not write to memory that outlives itself
+  // (potentially propagating derivative information), the only way to propagate
+  // derivative information is through the return value
+  // TODO the "doesn't write to active memory" can be made more aggressive than
+  // doesn't write to any memory
+  bool noActiveWrite = false;
+
+  if (isa<MemoryEffectOpInterface>(I) && !hasEffect<MemoryEffects::Write>(I))
+    noActiveWrite = true;
+  else if (auto CI = dyn_cast<CallOpInterface>(I)) {
+    // if (AA.onlyReadsMemory(CI)) {
+    //   noActiveWrite = true;
+    // } else 
+    if (Operation *F = getFunctionFromCall(CI)) {
+      // if (isMemFreeLibMFunction(F->getName())) {
+      //   noActiveWrite = true;
+      // } else
+      StringRef Name = cast<SymbolOpInterface>(F).getName();
+      if (Name == "frexp" || Name == "frexpf" || Name == "frexpl") {
+        noActiveWrite = true;
+      }
+    }
+  }
+  if (noActiveWrite) {
+    // Even if returning a pointer, this instruction is considered inactive
+    // since the instruction doesn't prop gradients. Thus, so long as we don't
+    // return an object containing a float, this instruction is inactive
+    // if (!TR.intType(1, I, /*errifNotFound*/ false).isPossibleFloat()) {
+    //   if (EnzymePrintActivity)
+    //     llvm::errs()
+    //         << " constant instruction from known non-float non-writing "
+    //            "instruction "
+    //         << *I << "\n";
+    //   InsertConstantInstruction(TR, I);
+    //   return true;
+    // }
+
+    // If all returned values constant otherwise, the operation is inactive
+    if (llvm::all_of(I->getResults(),
+                     [&](Value v) { return isConstantValue(TR, v); })) {
+      // if (EnzymePrintActivity)
+      //   llvm::errs() << " constant instruction from known constant non-writing "
+      //                   "instruction "
+      //                << *I << "\n";
+      InsertConstantOperation(TR, I);
+      return true;
+    }
+
+    // Even if the return is nonconstant, it's worth checking explicitly the
+    // users since unlike isConstantValue, returning a pointer does not make the
+    // instruction active
+    if (directions & DOWN) {
+      // We shall now induct on this instruction being inactive and try to prove
+      // this fact from a lack of active users.
+
+      // If we aren't a phi node (and thus potentially recursive on uses) and
+      // already equal to the current direction, we don't need to induct,
+      // reducing runtime.
+      if (directions == DOWN /*&& !isa<PHINode>(I)*/) {
+        if (llvm::all_of(I->getResults(), [&](Value val) {
```

Comment:
A lot of this code is written with the LLVM IR object inheritance structure in mind. Specifically, that PHI nodes are instructions and that instructions are values. This isn't true in MLIR. A more careful handling is necessary for block arguments, but it could actually simplify the code as a result. Checks for the instruction-as-value being constant are replaced with checks that _all_ results of an operation being constant.